R/harmonic.R

```diff
@@ -0,0 +1,374 @@
+  harmonic_data <- info[info$variable %in% col_names, ]
+
+  if (any(harmonic_data$type != "date" & harmonic_data$type != "numeric"))
+    rlang::abort(
+      paste0(
+        "All variables for `step_harmonic` should be either `Date` ",
+        "`POSIXct` or `numeric` classes."
+      )
+    )
+
+  # record the starting value for consistent phases
+  x$object <- lapply(training[, col_names], process_inputs,
+                           period = x$period,
+                           frequency = x$frequency,
+                           starting_val = x$starting_val,
+                           cycle_unit = x$cycle_unit)
+
+  names(x$object) <- col_names
+
+
+  step_harmonic_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    frequency = x$frequency,
+    period = x$period,
+    cycle_unit = x$cycle_unit,
+    starting_val = x$object$starting_val,
+    object = x$object,
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+
+sin_cos <- function(x,
+                    frequency,
+                    starting_val,
+                    cycle_size = 86400) {
+
+  if (all(is.na(x)))
+    rlang::abort("variable must have at least one non-NA value")
+
+  nc <- length(frequency)
+  nr <- length(x)
+
+  x <- x - as.numeric(starting_val)
+
+  # cycles per unit
+  cycle <- 2 * (pi * (x / cycle_size));
+
+  m <- matrix(NA_real_,
+         ncol = nc * 2,
+         nrow = nr)
+
+  for(i in seq_along(frequency)) {
+    m[, i] <- sin(cycle * frequency[i])
+    m[, i + nc] <- cos(cycle * frequency[i])
+  }
+
+
+  return(m)
+}
+
+
+
+#' @export
+bake.step_harmonic <- function(object, new_data, ...) {
+
+  for (col_name in names(object$objects)) {
+    ob <- object$objects[[col_name]]
+    res <- sin_cos(as.numeric(new_data[[col_name]]),
+                   ob[["frequency"]],
+                   ob[["starting_val"]],
+                   ob[["cycle_size"]])
+    colnames(res) <- paste0(col_name, '_', ob[["col_names"]])
+    res <- as_tibble(res)
+    new_data <- bind_cols(new_data, res)
+
+  }
+
+
+  as_tibble(new_data)
+}
+
+#' @export
+print.step_harmonic <-
+  function(x, width = max(20, options()$width - 30), ...) {
+    cat("Harmonic numeric variables for ", sep = "")
+    printer(tr_obj = names(x$object), untr_obj = x$terms,
+            trained = x$trained, width = width)
+    invisible(x)
+  }
+
+
+#' @rdname tidy.recipe
+#' @param x A `step_harmonic` object.
+#' @export
+tidy.step_harmonic <- function(x, ...) {
+  if (is_trained(x)) {
+    res <-
+      tibble(terms = names(x$objects),
+             value = sapply(x$objects, function(y) y$starting_val))
+  } else {
+    term_names <- sel2char(x$terms)
+    res <- tibble(terms = term_names,
+                  value = na_dbl)
+  }
+  res$id <- x$id
+  res
+}
+
+
+# may want a tunable step that can adjust the frequency/period
+#' #' @rdname tunable.step
+#' #' @export
+#' tunable.step_harmonic <- function(x, ...) {
+#'   tibble::tibble(
+#'     name = c("frequency", "period"),
```

Comment:
`cycle_unit` too?

Reply:
They shouldn't tune over this or over `cycle_size`.  Now tuning over only `frequency` should be preferred all that is needed.